### PR TITLE
Refactors StackdriverTracePropagation so that B3 can be reconfigured

### PR DESCRIPTION
**Pull request by adriancole**
For a few releases, Brave has been able to configure `B3` differently
per request type. For example, messaging configured with single headers.

This makes it possible for `StackdriverTracePropagation` to honor that,
notably by fixing an incorrect package name! For example, depending on
the version of sleuth in use, this can now change the policy for which
headers are used for local propagation.

See https://github.com/spring-cloud/spring-cloud-sleuth/issues/1643

### CODE REVIEW

File: propagation-stackdriver/README.md [hunk 1]
```diff
@@ -4,19 +4,19 @@ This is a propagation that behaves like `B3Propagation`, but falls back on traci
...
 
-This propagation makes use of the `CompositeExtractor` concept, which attempts to extract the tracing context from multiple extractors.
-At the moment, it doesn't support partial extraction, meaning that if one extractor returns a context that is not empty, that context is returned and the following ones are not ran.
-
-The first attempted extractor is the `B3Propagation` one, using the `X-B3-TraceId`, `X-B3-SpanId`, etc. keys.
-
-If `B3Propagation` can't find a context, the next extractor is the `XCloudTraceContextExtractor`.
-It checks the `x-cloud-trace-context` key, which is structured in the following way:
+For example, if the "b3" header exists, it is used and any "x-cloud-trace-context" will be ignored.
+Whyen absent, the following applies:
```
Comment (adriancole) [inline propagation-stackdriver/README.md:19]:
```suggestion
When absent, the following applies:
```

**Comment by adriancole** [comment]
shade looks correct:

```
± |propagation-redo ?:3 ✗| → jar -tvf /Users/acole/.m2/repository/io/zipkin/gcp/brave-propagation-stackdriver/0.16.2-SNAPSHOT/brave-propagation-stackdriver-0.16.2-SNAPSHOT.jar
   135 Mon May 18 16:12:18 MYT 2020 META-INF/MANIFEST.MF
     0 Mon May 18 16:12:18 MYT 2020 META-INF/
     0 Mon May 18 16:12:16 MYT 2020 zipkin2/
     0 Mon May 18 16:12:16 MYT 2020 zipkin2/propagation/
     0 Mon May 18 16:12:16 MYT 2020 zipkin2/propagation/stackdriver/
  1992 Mon May 18 16:12:16 MYT 2020 zipkin2/propagation/stackdriver/StackdriverTracePropagation$1.class
  3403 Mon May 18 16:12:16 MYT 2020 zipkin2/propagation/stackdriver/StackdriverTracePropagation.class
     0 Mon May 18 16:12:16 MYT 2020 brave/
     0 Mon May 18 16:12:16 MYT 2020 brave/propagation/
     0 Mon May 18 16:12:16 MYT 2020 brave/propagation/stackdriver/
  2092 Mon May 18 16:12:16 MYT 2020 brave/propagation/stackdriver/StackdriverTracePropagation$Factory.class
  3392 Mon May 18 16:12:16 MYT 2020 brave/propagation/stackdriver/StackdriverTracePropagation.class
  6809 Mon May 18 16:12:16 MYT 2020 brave/propagation/stackdriver/XCloudTraceContextExtractor.class
     0 Mon May 18 15:58:02 MYT 2020 META-INF/maven/
     0 Mon May 18 15:58:02 MYT 2020 META-INF/maven/io.zipkin.gcp/
     0 Mon May 18 15:58:02 MYT 2020 META-INF/maven/io.zipkin.gcp/brave-propagation-stackdriver/
  3067 Mon May 18 15:58:02 MYT 2020 META-INF/maven/io.zipkin.gcp/brave-propagation-stackdriver/pom.xml
    87 Mon May 18 16:12:18 MYT 2020 META-INF/maven/io.zipkin.gcp/brave-propagation-stackdriver/pom.properties
     0 Sun May 17 10:52:54 MYT 2020 brave/propagation/stackdriver/internal/
     0 Sun May 17 10:52:54 MYT 2020 brave/propagation/stackdriver/internal/propagation/
  2584 Sun May 17 10:52:54 MYT 2020 brave/propagation/stackdriver/internal/propagation/StringPropagationAdapter$GetterAdapter.class
  2619 Sun May 17 10:52:54 MYT 2020 brave/propagation/stackdriver/internal/propagation/StringPropagationAdapter$SetterAdapter.class
  5373 Sun May 17 10:52:54 MYT 2020 brave/propagation/stackdriver/internal/propagation/StringPropagationAdapter.class
```

File: propagation-stackdriver/pom.xml [hunk 2]
```diff
@@ -25,13 +25,63 @@
...
+                <includes>
+                  <include>${brave.groupId}:brave</include>
+                </includes>
+              </artifactSet>
+              <filters>
+                <filter>
+                  <artifact>${brave.groupId}:brave</artifact>
+                  <includes>
+                    <include>brave/internal/propagation/StringPropagationAdapter*.class</include>
+                    <include>brave/internal/codec/HexCodec*.class</include>
```
Comment (adriancole) [inline propagation-stackdriver/pom.xml:64]: @elefeint no more bomb :D

**Comment by adriancole** [comment]
jar now also has HexCodec in it

```
± |propagation-redo ?:3 ✗| → jar -tvf ./propagation-stackdriver/target/brave-propagation-stackdriver-0.16.2-SNAPSHOT.jar
   135 Tue May 19 08:38:34 MYT 2020 META-INF/MANIFEST.MF
     0 Tue May 19 08:38:34 MYT 2020 META-INF/
     0 Tue May 19 08:38:32 MYT 2020 zipkin2/
     0 Tue May 19 08:38:32 MYT 2020 zipkin2/propagation/
     0 Tue May 19 08:38:32 MYT 2020 zipkin2/propagation/stackdriver/
  1992 Tue May 19 08:38:32 MYT 2020 zipkin2/propagation/stackdriver/StackdriverTracePropagation$1.class
  3403 Tue May 19 08:38:32 MYT 2020 zipkin2/propagation/stackdriver/StackdriverTracePropagation.class
     0 Tue May 19 08:38:32 MYT 2020 brave/
     0 Tue May 19 08:38:32 MYT 2020 brave/propagation/
     0 Tue May 19 08:38:32 MYT 2020 brave/propagation/stackdriver/
  2092 Tue May 19 08:38:32 MYT 2020 brave/propagation/stackdriver/StackdriverTracePropagation$Factory.class
  3392 Tue May 19 08:38:32 MYT 2020 brave/propagation/stackdriver/StackdriverTracePropagation.class
  6809 Tue May 19 08:38:32 MYT 2020 brave/propagation/stackdriver/XCloudTraceContextExtractor.class
     0 Tue May 19 08:31:52 MYT 2020 META-INF/maven/
     0 Tue May 19 08:31:52 MYT 2020 META-INF/maven/io.zipkin.gcp/
     0 Tue May 19 08:31:52 MYT 2020 META-INF/maven/io.zipkin.gcp/brave-propagation-stackdriver/
  3143 Tue May 19 08:31:52 MYT 2020 META-INF/maven/io.zipkin.gcp/brave-propagation-stackdriver/pom.xml
    87 Tue May 19 08:38:34 MYT 2020 META-INF/maven/io.zipkin.gcp/brave-propagation-stackdriver/pom.properties
     0 Mon May 18 12:30:26 MYT 2020 brave/propagation/stackdriver/internal/
     0 Mon May 18 12:30:26 MYT 2020 brave/propagation/stackdriver/internal/codec/
  2733 Mon May 18 12:30:26 MYT 2020 brave/propagation/stackdriver/internal/codec/HexCodec.class
     0 Mon May 18 12:30:26 MYT 2020 brave/propagation/stackdriver/internal/propagation/
  2584 Mon May 18 12:30:26 MYT 2020 brave/propagation/stackdriver/internal/propagation/StringPropagationAdapter$GetterAdapter.class
  2619 Mon May 18 12:30:26 MYT 2020 brave/propagation/stackdriver/internal/propagation/StringPropagationAdapter$SetterAdapter.class
  5373 Mon May 18 12:30:26 MYT 2020 brave/propagation/stackdriver/internal/propagation/StringPropagationAdapter.class
```

File: propagation-stackdriver/pom.xml (hunk 2 above)
Comment (elefeint) [reply propagation-stackdriver/pom.xml:64]: Thank you! This is great; I did not consider shading as a solution; I kept trying to figure out how to enforce version uniformity in spring-cloud-gcp.

Comment (adriancole) [reply propagation-stackdriver/pom.xml:64]: some internal things aren't written to be shaded well, these were.. sometimes you can lose time in the shade :D

Comment (adriancole) [reply propagation-stackdriver/pom.xml:64]: ex anything that references a thread local can't be shaded mostly

**Pull request merged**